### PR TITLE
Bundle standard contracts with MIRAI binary

### LINKED ISSUE #2: Figure out how to build summaries for the rust standard library
Building the standard library is somewhat special. Here are instructions from the Miri project:

```
rustup component add --toolchain nightly rust-src
cargo +nightly install xargo
rustup run nightly xargo/build.sh
```

One of the first things to figure out is how to run Mirai over the standard library code and to output summaries that can be used in the same way as summaries for any other crate.

### PR DESCRIPTION
**Pull request by hermanventer**
## Description

An as yet minimal set of standard (core) library function summaries are now included in the MIRAI binary. When MIRAI analyses a project that does not have an existing summary store, MIRAI will not just create an empty store, but also pre-populate it with summaries for (some) standard functions.

For the time being, summaries will be added as needed. Eventually there will have to be a way to compute them directly from the standard library sources.

To add a summary, edit the standard_contracts/src/foreign_contracts.cs (or eventually the appropriate sub module file) and add code that will result in the desired summary. Then run the rebuild_std.sh script in the root directory.

When testing, make sure that the output directory does not contain a summary store (or a tar ball of a summary store).

Note that (most) test files no longer include foreign_contracts modules since these are subsumed by the standard summaries.

Fixes #2 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] API change with a documentation update
- [ ] Additional test coverage
- [ ] Code cleanup or just keeping up with the latest Rustc nightly

## How Has This Been Tested?
cargo test; ./validate.sh

### CODE REVIEW

File: checker/tests/run-pass/for_iterator.rs [hunk 1]
```diff
@@ -1,91 +1,67 @@
...
-                    value.record_heap_addresses(&mut new_roots);
-                    result.push((path.clone(), value.clone()));
-                }
-            }
-        }
-        heap_roots.extend(new_roots.into_iter());
-    }
-    result
-}
+//use std::collections::{HashMap, HashSet};
```
Comment (hermanventer) [inline checker/tests/run-pass/for_iterator.rs]: This test, when run with non default (empty) summaries, triggers an exponential blow up. Fixing this is out of scope for this PR.

File: standard_contracts/README.md [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# MIRAI Standard Contracts
+
+This crate provides a set of function definitions that provide summaries of parts of the RUST standard library.
```
Comment (darioncassel) [inline standard_contracts/README.md]: Is Rust in uppercase for emphasis? I believe Rust is capitalized as "Rust".

File: standard_contracts/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,17 @@
+[package]
+
+name = "mirai-standard-contracts"
+version = "0.0.1"
+authors = ["Herman Venter <hermanv@fb.com>"]
+description = "Specifications of standard library functions for MIRAI"
+repository = "https://github.com/facebookexperimental/MIRAI"
+readme = "README.md"
+license = "MIT"
+edition = "2018"
```
Comment (darioncassel) [inline standard_contracts/Cargo.toml]: Should the edition be 2019 now?

File: checker/src/summaries.rs [hunk 4]
```diff
@@ -266,6 +269,37 @@ impl<'a, 'tcx: 'a> PersistentSummaryCache<'a, 'tcx> {
...
+    ) -> std::path::PathBuf {
+        use std::env;
+        use std::fs::File;
+        use std::io::Write;
+        use std::path::Path;
+        use tar::Archive;
+
+        let directory_path = Path::new(summary_store_directory_str);
+        let store_path = directory_path.join(".summary_store.sled");
+        if !store_path.exists() && env::var("MIRAI_START_FRESH").is_err() {
```
Comment (darioncassel) [inline checker/src/summaries.rs]: (Out of scope for this PR: Perhaps we should have all of the MIRAI flags documented somewhere?)

File: standard_contracts/Cargo.toml (hunk 3 above)
Comment (hermanventer) [reply standard_contracts/Cargo.toml]: The edition refers to the version of Rust that must be used to compile this. 2018 still seems appropriate.

File: checker/src/summaries.rs (hunk 4 above)
Comment (hermanventer) [reply checker/src/summaries.rs]: Good idea.
